src/protocols/secure_channel/tests/TestDefaultSessionResumptionStorage.cpp

```diff
@@ -81,7 +83,124 @@ void TestSave(nlTestSuite * inSuite, void * inContext)
     }
 
     // Read back and verify values.
-    for (size_t i = 0; i < CHIP_CONFIG_CASE_SESSION_RESUME_CACHE_SIZE; ++i)
+    for (auto & vector : vectors)
+    {
+        chip::ScopedNodeId outNode;
+        chip::SessionResumptionStorage::ResumptionIdStorage outResumptionId;
+        chip::Crypto::P256ECDHDerivedSecret outSharedSecret;
+        chip::CATValues outCats;
+
+        // Verify retrieval by node.
+        NL_TEST_ASSERT(inSuite,
+                       sessionStorage.FindByScopedNodeId(vector.node, outResumptionId, outSharedSecret, outCats) == CHIP_NO_ERROR);
+        NL_TEST_ASSERT(inSuite, memcmp(vector.resumptionId.data(), outResumptionId.data(), vector.resumptionId.size()) == 0);
+        NL_TEST_ASSERT(inSuite, memcmp(vector.sharedSecret.ConstBytes(), outSharedSecret, vector.sharedSecret.Length()) == 0);
+        NL_TEST_ASSERT(inSuite, vector.cats.values[0] == outCats.values[0]);
+        NL_TEST_ASSERT(inSuite, vector.cats.values[1] == outCats.values[1]);
+        NL_TEST_ASSERT(inSuite, vector.cats.values[2] == outCats.values[2]);
+
+        // Validate retreiveal by resumption ID.
+        NL_TEST_ASSERT(inSuite,
+                       sessionStorage.FindByResumptionId(vector.resumptionId, outNode, outSharedSecret, outCats) == CHIP_NO_ERROR);
+        NL_TEST_ASSERT(inSuite, vector.node == outNode);
+        NL_TEST_ASSERT(inSuite, memcmp(vector.sharedSecret.Bytes(), outSharedSecret, vector.sharedSecret.Length()) == 0);
+        NL_TEST_ASSERT(inSuite, vector.cats.values[0] == outCats.values[0]);
+        NL_TEST_ASSERT(inSuite, vector.cats.values[1] == outCats.values[1]);
+        NL_TEST_ASSERT(inSuite, vector.cats.values[2] == outCats.values[2]);
+    }
+}
+
+void TestInPlaceSave(nlTestSuite * inSuite, void * inContext)
+{
+    chip::SimpleSessionResumptionStorage sessionStorage;
+    chip::TestPersistentStorageDelegate storage;
+    sessionStorage.Init(&storage);
+    struct
+    {
+        chip::SessionResumptionStorage::ResumptionIdStorage resumptionId;
+        chip::Crypto::P256ECDHDerivedSecret sharedSecret;
+        chip::ScopedNodeId node;
+        chip::CATValues cats;
+    } vectors[CHIP_CONFIG_CASE_SESSION_RESUME_CACHE_SIZE + 10];
+
+    // Construct only a few unique node identities to simiulate talking to a
+    // couple peers.
+    chip::ScopedNodeId nodes[3];
+    static_assert(ARRAY_SIZE(nodes) < CHIP_CONFIG_CASE_SESSION_RESUME_CACHE_SIZE,
+                  "must have fewer nodes than slots in session resumption storage");
+    for (size_t i = 0; i < ARRAY_SIZE(nodes); ++i)
+    {
+        do
+        {
+            nodes[i] = chip::ScopedNodeId(static_cast<chip::NodeId>(rand()), static_cast<chip::FabricIndex>(i + 1));
+        } while (!nodes[i].IsOperational());
+    }
+
+    // Populate test vectors.
+    for (size_t i = 0; i < ARRAY_SIZE(vectors); ++i)
```

Comment:
A lot of these loops that just use `i` to index into `vectors` or `nodes` might have been clearer as ranged for loops....

Reply:
I think in all cases where there are indexed loops, the index variable is being used to populate the test vectors with values that are guaranteed not to collide.  I tried to move to ranged for loops wherever possible.

Reply:
Ah, I missed some `i` on the rhs of things, indeed.